### ISSUE TITLE
After upgrading to Visual Studio Code 1.92.0, Source Control tab peformance is unusable due to new showHistoryGraph feature.

### ISSUE THREAD

**Issue by mazer1310** [microsoft/vscode#224646]
Type: <b>Bug</b>

After upgrading to Visual Studio Code 1.92.0, the Source Control tab does not refresh for 5 to 10 minutes.

Looking in Process Explorer, there are multiple hanging git processes of this form:

git.exe log --format=%H%n%aN%n%aE%n%at%n%ct%n%P%n%D%n%B -z --shortstat --diff-merges=first-parent 1a766051b00f4f633309d9a9f8c86b4492b9f664.. --topo-order --decorate=full refs/heads/develop/pc_d refs/remotes/origin/develop/pc_d refs/remotes/origin/main

each of which run for 5-10 minutes before finally exiting to be immediately replaced with another similar git process

After 5 to 10 minutes, the tab eventually appears, but any change to any file causes the tab to freeze again for a similar period of time.  As a result, the Source Control tab is completely unusable.  This was not a problem in 1.91.3.

After looking at the 1.92.0 Release Notes, I disabled the "scm.showHistoryGraph" setting and the control is back to being usable again, so it appears the performance of the new feature is not usable for our repository.

VS Code version: Code 1.92.0 (b1c0a14de1414fcdaa400695b4db1c0799bc3124, 2024-07-31T23:26:45.634Z)
OS version: Windows_NT x64 10.0.22621
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz (12 x 2592)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|31.78GB (17.87GB free)|
|Process Argv|--crash-reporter-id 4e97402e-edf9-4012-b7a5-1934ce38844d|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (23)</summary>

Extension|Author (truncated)|Version
---|---|---
githistory|don|0.6.20
gitlens|eam|15.2.3
prettier-vscode|esb|10.4.0
git-project-manager|fel|1.8.2
vscode-solution-explorer|fer|0.8.6
gitlab-workflow|Git|5.3.0
git-worktrees|Git|2.2.0
compare-helper|kee|1.1.1
git-tree-compare|let|1.16.0
git-graph|mhu|1.30.0
debugpy|ms-|2024.10.0
python|ms-|2024.12.1
vscode-pylance|ms-|2024.7.1
remote-ssh|ms-|0.112.0
remote-ssh-edit|ms-|0.86.0
cmake-tools|ms-|1.18.44
cpptools|ms-|1.21.5
cpptools-extension-pack|ms-|1.3.0
remote-explorer|ms-|0.4.3
markdown-preview-enhanced|shd|0.8.13
cmake|twx|0.0.17
gitblame|wad|11.0.1
markdown-all-in-one|yzh|3.6.2

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscoreces:30445986
vscod805:30301674
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
vscrp:30673768
962ge761:30959799
pythongtdpath:30769146
welcomedialogc:30910334
pythonnoceb:30805159
asynctok:30898717
pythonregdiag2:30936856
pythonmypyd1:30879173
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
0ee40948:31013168
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31068756
dwnewjupytercf:31046870
newcmakeconfigv2:31071590
impr_priority:31102340
nativerepl1:31104043
refactort:31108082
ccpltc:31103427
pythonrstrctxt:31103193

```

</details>

**Comment by pitaj** [comment]
Same issue here, but it was also causing VS Code to hang eventually after opening Source Control in the sidebar. Disabling the setting fixed it.

`git version 2.45.2`

<details>
<summary>System Info</summary>

```text
Version: 1.92.0
Commit: b1c0a14de1414fcdaa400695b4db1c0799bc3124
Date: 2024-07-31T23:26:45.634Z
Electron: 30.1.2
ElectronBuildId: 9870757
Chromium: 124.0.6367.243
Node.js: 20.14.0
V8: 12.4.254.20-electron.0
OS: Linux x64 6.6.41-1-MANJARO
```

</details>

**Comment by arvl130** [comment]
Same issue on Arch Linux x86_64. `git` is using up a lot of CPU trying to generate the list of commits. Likewise, it looks like Electron is having a hard time rendering the graph.

From what I see, the issue is with the graph itself. It tries to render a graph starting from the ***very first*** commit. This maybe intended behavior, but it causes a performance issue for repositories that already has 15,000+ commits like ours.

<details>
    <summary>System Info</summary>

    Version: 1.92.0
    Commit: b1c0a14de1414fcdaa400695b4db1c0799bc3124
    Date: 2024-08-01T22:07:15.925Z
    Electron: 30.3.1
    ElectronBuildId: undefined
    Chromium: 124.0.6367.243
    Node.js: 20.15.1
    V8: 12.4.254.20-electron.0
    OS: Linux x64 6.10.2-arch1-2

</details>

**Comment by pitaj** [comment]
That makes sense. I was working on rust-lang/rust.

**Comment by irfan7junior** [comment]
> Same issue here, but it was also causing VS Code to hang eventually after opening Source Control in the sidebar. Disabling the setting fixed it.
> 
> `git version 2.45.2`
> 
> System Info
> ```
> Version: 1.92.0
> Commit: b1c0a14de1414fcdaa400695b4db1c0799bc3124
> Date: 2024-07-31T23:26:45.634Z
> Electron: 30.1.2
> ElectronBuildId: 9870757
> Chromium: 124.0.6367.243
> Node.js: 20.14.0
> V8: 12.4.254.20-electron.0
> OS: Linux x64 6.6.41-1-MANJARO
> ```

can you help with the setting name? I want to disable it too.

**Comment by gjsjohnmurray** [comment]
> can you help with the setting name? I want to disable it too.

https://github.com/microsoft/vscode/issues/224768#issuecomment-2268966616

I recommend reviewing the Release Notes after every VS Code update.

**Comment by promethyttrium** [comment]
Or you can default features that will create performance degradation (like a giant history graph) to off and let use decide if we want the feature or not.

**Comment by JCEleanor** [comment]
TLDR: add this to your `setting.json`
` "scm.showHistoryGraph": false`

**Comment by JasonWHowell** [comment]
Related
https://github.com/microsoft/vscode/issues/225153
https://github.com/microsoft/vscode/issues/225210
https://github.com/microsoft/vscode/issues/225088
https://github.com/microsoft/vscode/issues/225006
https://github.com/microsoft/vscode/issues/224916
https://github.com/microsoft/vscode/issues/224646

**Comment by rmcvey** [comment]
FWIW I only really needed to set `scm.showOutgoingChanges` to "never" for Changes to be reflected again (slower than normal). However, disabling `showHistoryGraph` AND `showOutgoingChanges` made it work as expected.

```
  "scm.showHistoryGraph": false,
  "scm.showOutgoingChanges": "never",
```

**Comment by joelliang** [comment]
Same issue here, I tried to run the "git log" command in the command line, the command took a long time. But it's much faster if I remove the "--shortstat" from the command line argument.

**Comment by lszomoru** [comment]
The history graph has been separated out in a new view called "Source Control Graph". The new view implements both caching as well as paging so the performance issue should be resolved. If the view is collapsed on hidden it will never query for the repository history. The change is already available in VS Code Insiders and it will be included in the next VS Code Stable release.
